File: mynlp-segment/src/main/java/com/mayabot/nlp/segment/tokenizer/collector/SentenceCollector.java
package com.mayabot.nlp.segment.tokenizer.collector;

import com.mayabot.nlp.segment.WordTerm;
import com.mayabot.nlp.segment.WordTermCollector;
import com.mayabot.nlp.segment.wordnet.Vertex;
import com.mayabot.nlp.segment.wordnet.Wordnet;
import com.mayabot.nlp.segment.wordnet.Wordpath;
import com.mayabot.nlp.utils.StringUtils;

import java.util.Iterator;
import java.util.function.Consumer;

/**
 * Nlp收集方式，不处理子词
 * 按照WordPath里面描述的唯一切分路径，构建WordTerm序列
 */
public class SentenceCollector implements WordTermCollector {

    @Override
    public void collect(Wordnet wordnet, Wordpath wordPath, Consumer<WordTerm> consumer) {
        Iterator<Vertex> vertexIterator = wordPath.iteratorVertex();
        while (vertexIterator.hasNext()) {
            Vertex vertex = vertexIterator.next();

            WordTerm term = new WordTerm(vertex.realWord(), vertex.guessNature());
            term.setOffset(vertex.getRowNum());

            //默认把空白的字符去除掉
            if (StringUtils.isWhiteSpace(term.word)) {
                continue;
            }
        }
    }
}

File: mynlp/src/main/java/com/mayabot/nlp/segment/plugins/collector/SentenceCollector.kt
package com.mayabot.nlp.segment.plugins.collector

import com.mayabot.nlp.Mynlp
import com.mayabot.nlp.common.utils.StringUtils
import com.mayabot.nlp.segment.WordTerm
import com.mayabot.nlp.segment.wordnet.Wordnet
import com.mayabot.nlp.segment.wordnet.Wordpath
import java.util.function.Consumer

/**
 * WordTermCollector的默认实现，从各种数据结构中收集和生成词序列
 *
 * @author jimichan
 */
class SentenceCollector(
    private val mynlp: Mynlp,
    private val subwordComputer: List<SubwordComputer> = emptyList(),
    private val setupList: List<SubwordInfoSetup> = emptyList()
) : WordTermCollector {

    override fun collect(txtChars: CharArray?, wordnet: Wordnet, wordPath: Wordpath, consumer: Consumer<WordTerm>) {

        val vertexIterator = wordPath.iteratorVertex()

        setupList.forEach {
            it.fill(wordnet, wordPath)
        }

        while (vertexIterator.hasNext()) {
            val vertex = vertexIterator.next()

            val word = if (txtChars == null) {
                vertex.realWord()
            } else {
                String(chars = txtChars, vertex.offset(), vertex.length)
            }

            val term = WordTerm(word, vertex.nature, vertex.offset())

            if (StringUtils.isWhiteSpace(term.word)) {
                continue
            }

            // 如果运行成功，后面的就不运行了
            subwordComputer.forEach {
                if (it.run(term, wordnet, wordPath)) {
                    return@forEach
                }
            }

            consumer.accept(term)
        }
    }
}